File: packages/contracts/src/audit.ts
import type { Actor, ID, ISODateTime, Role } from "./global";

export type AuditAction =
  | "TASK_CREATED"
  | "TASK_PUBLISHED"
  | "TASK_PAUSED"
  | "TASK_RESUMED"
  | "TASK_ENDED"
  | "TASK_ARCHIVED"
  | "SCHEMA_DRAFT_SAVED"
  | "SCHEMA_VERSION_PUBLISHED"
  | "DATASET_IMPORTED"
  | "ASSIGNMENT_CLAIMED"
  | "ASSIGNMENT_EXPIRED"
  | "DRAFT_SAVED"
  | "SUBMISSION_CREATED"
  | "AI_REVIEW_ENQUEUED"
  | "AI_REVIEW_STARTED"
  | "AI_REVIEW_SUCCEEDED"
  | "AI_REVIEW_FAILED"
  | "AI_REVIEW_FAILED_TO_HUMAN"
  | "REVIEW_CLAIMED"
  | "FINAL_REVIEW_REQUESTED"
  | "REVIEW_RETURNED"
  | "REVIEW_ACCEPTED"
  | "REVIEW_REJECTED"
  | "EXPORT_CREATED"
  | "EXPORT_STARTED"
  | "EXPORT_SUCCEEDED"
  | "EXPORT_FAILED"
  | "EXPORT_CANCELED"
  | "FILE_UPLOAD_URL_CREATED"
  | "FILE_UPLOAD_STARTED"
  | "FILE_UPLOAD_FAILED"
  | "FILE_UPLOADED"
  | "FILE_CONFIRMED";

export interface AuditLog {
  id: ID;
  entityType:
    | "TASK"
    | "SCHEMA"
    | "ITEM"
    | "ASSIGNMENT"
    | "SUBMISSION"
    | "REVIEW"
    | "AI_REVIEW_JOB"
    | "EXPORT"
    | "FILE";
  entityId: ID;
  action: AuditAction;
  actor: Actor;
  before?: unknown;
  after?: unknown;
  reason?: string;
  requestId?: string;
  createdAt: ISODateTime;
}

export interface AuditLogSummary {
  id: ID;
  action: AuditAction;
  createdAt: ISODateTime;
}

export type AuditEventType =
  | AuditAction
  | "SCHEMA_PUBLISH_REQUESTED"
  | "SCHEMA_COMPATIBILITY_CHECKED"
  | "SCHEMA_PUBLISH_BLOCKED"
  | "SCHEMA_PUBLISH_FAILED"
  | "DEPRECATION_WARNING_GENERATED"
  | "MIGRATION_PLAN_CREATED"
  | "MIGRATION_DRY_RUN_COMPLETED"
  | "MIGRATION_EXECUTED"
  | "SUBMISSION_UPDATED"
  | "REVIEW_SUBMITTED"
  | "AI_REVIEW_GENERATED"
  | "EXPORT_GENERATED";

export type AuditSeverity = "INFO" | "WARNING" | "ERROR";

export type AuditSource =
  | "WEB"
  | "API"
  | "WORKER"
  | "AI_AGENT"
  | "SYSTEM";

export type AuditActor = {
  id: ID | string;
  role: Role | string;
  displayName?: string;
};

export type AuditTargetEntityType =
  | "TASK"
  | "SCHEMA"
  | "SCHEMA_VERSION"
  | "MIGRATION"
  | "ITEM"
  | "ASSIGNMENT"
  | "SUBMISSION"
  | "REVIEW"
  | "AI_REVIEW_JOB"
  | "EXPORT"
  | "FILE";

export type AuditTarget = {
  entityType: AuditTargetEntityType;
  entityId: ID | string;
  taskId?: ID | string;
  schemaId?: ID | string;
  schemaVersionId?: ID | string;
  assignmentId?: ID | string;
  submissionId?: ID | string;
  reviewId?: ID | string;
  exportId?: ID | string;
  migrationPlanId?: ID | string;
};

export type SchemaDraftSavedAuditPayload = {
  schemaDraftRevision?: number;
  fieldCount?: number;
  validationErrorCount?: number;
  validationWarningCount?: number;
};

export type SchemaPublishRequestedAuditPayload = {
  schemaDraftRevision?: number;
  isFirstPublish: boolean;
  requiresApproval: boolean;
  requiresMigration: boolean;
  confirmedByActor?: boolean;
};

export type SchemaCompatibilityCheckedAuditPayload = {
  compatible: boolean;
  publishAllowed: boolean;
  requiresApproval: boolean;
  requiresMigration: boolean;
  changeCodes: string[];
  blockingCount: number;
  warningCount: number;
  reportChecksum?: string;
};

export type SchemaPublishBlockedAuditPayload = {
  blockingChangeCodes: string[];
  deprecationErrorCodes?: string[];
  schemaValidationErrorCount?: number;
};

export type SchemaPublishFailedAuditPayload = {
  reasonCode?: string;
  message: string;
  stage?: "SAVE_DRAFT" | "PUBLISH_SCHEMA" | "PUBLISH_TASK" | "UNKNOWN";
};

export type SchemaPublishedAuditPayload = {
  schemaVersionId: ID | string;
  schemaVersionNo?: number;
  snapshotHash?: string;
};

export type DeprecationWarningGeneratedAuditPayload = {
  warningCodes: string[];
  fieldNames?: string[];
  warningCount: number;
};

export type MigrationPlanCreatedAuditPayload = {
  fromSchemaVersionId?: ID | string;
  toSchemaVersionId?: ID | string;
  operationCount: number;
  manualMappingSlotCount: number;
  executable: boolean;
  blockingCount: number;
  planChecksum?: string;
};

export type MigrationDryRunCompletedAuditPayload = {
  totalSubmissions: number;
  affectedSubmissions: number;
  validationErrorCount: number;
  skippedCount: number;
  sampleCount: number;
  reportChecksum?: string;
};

export type MigrationExecutedAuditPayload = {
  operationCount: number;
  migratedCount: number;
  skippedCount: number;
  conflictCount: number;
  executionChecksum?: string;
};

export type SubmissionAuditPayload = {
  schemaVersionId?: ID | string;
  answerHash?: string;
  changedFieldNames?: string[];
  validationErrorCount?: number;
  attemptNo?: number;
};

export type ReviewSubmittedAuditPayload = {
  decision: string;
  stage?: string;
  reasonCode?: string;
  patchedFieldNames?: string[];
  patchCount?: number;
};

export type AiReviewGeneratedAuditPayload = {
  aiReviewJobId?: ID | string;
  decision?: string;
  score?: number;
  confidence?: number;
  issueCount?: number;
  outputHash?: string;
  promptSnapshotHash?: string;
};

export type ExportGeneratedAuditPayload = {
  exportId?: ID | string;
  format?: string;
  rowCount?: number;
  warningCount?: number;
  mappingChecksum?: string;
};

export type GenericAuditEventPayload = {
  summary?: string;
  detailRef?: ID | string;
  counters?: Record<string, number>;
  codes?: string[];
};

export type AuditEventPayload =
  | SchemaDraftSavedAuditPayload
  | SchemaPublishRequestedAuditPayload
  | SchemaCompatibilityCheckedAuditPayload
  | SchemaPublishBlockedAuditPayload
  | SchemaPublishFailedAuditPayload
  | SchemaPublishedAuditPayload
  | DeprecationWarningGeneratedAuditPayload
  | MigrationPlanCreatedAuditPayload
  | MigrationDryRunCompletedAuditPayload
  | MigrationExecutedAuditPayload
  | SubmissionAuditPayload
  | ReviewSubmittedAuditPayload
  | AiReviewGeneratedAuditPayload
  | ExportGeneratedAuditPayload
  | GenericAuditEventPayload;

export type AuditEventRecord = {
  id: ID | string;
  type: AuditEventType;
  severity: AuditSeverity;
  source: AuditSource;
  actor: AuditActor;
  target: AuditTarget;
  payload: AuditEventPayload;
  requestId?: string;
  idempotencyKey?: string;
  checksum?: string;
  createdAt: ISODateTime | string;
};

export type AuditEventQuery = {
  taskId?: ID | string;
  entityType?: AuditTargetEntityType;
  entityId?: ID | string;
  schemaVersionId?: ID | string;
  assignmentId?: ID | string;
  submissionId?: ID | string;
  reviewId?: ID | string;
  exportId?: ID | string;
  migrationPlanId?: ID | string;
  actorId?: ID | string;
  types?: AuditEventType[];
  severities?: AuditSeverity[];
  source?: AuditSource;
  createdFrom?: ISODateTime | string;
  createdTo?: ISODateTime | string;
  limit?: number;
};

export type AppendAuditEventRequest = {
  type: AuditEventType;
  severity?: AuditSeverity;
  source: AuditSource;
  actor: AuditActor;
  target: AuditTarget;
  payload: AuditEventPayload;
  requestId?: string;
  idempotencyKey?: string;
  checksum?: string;
};

File: apps/api/app/models/audit.py
# audit_logs 表 ORM 模型，对应契约 §25 AuditLog 与 §24 存储契约。
# ⚠️ 本表禁止物理删除，应用层必须拒绝所有 DELETE 操作（契约 §24 / §25 明确规定）。
# 本表为追加只写的不可变审计记录，无 updated_at 字段。
# action 合法值（契约 §25 AuditAction，最长值 AI_REVIEW_FAILED_TO_HUMAN = 25 字符，使用 String(40)）：
#   TASK_CREATED / TASK_PUBLISHED / TASK_PAUSED / TASK_RESUMED / TASK_ENDED / TASK_ARCHIVED /
#   SCHEMA_DRAFT_SAVED / SCHEMA_VERSION_PUBLISHED / DATASET_IMPORTED /
#   ASSIGNMENT_CLAIMED / ASSIGNMENT_EXPIRED / DRAFT_SAVED / SUBMISSION_CREATED /
#   AI_REVIEW_ENQUEUED / AI_REVIEW_STARTED / AI_REVIEW_SUCCEEDED /
#   AI_REVIEW_FAILED / AI_REVIEW_FAILED_TO_HUMAN /
#   REVIEW_CLAIMED / FINAL_REVIEW_REQUESTED / REVIEW_RETURNED / REVIEW_ACCEPTED / REVIEW_REJECTED /
#   EXPORT_CREATED / EXPORT_STARTED / EXPORT_SUCCEEDED / EXPORT_FAILED / EXPORT_CANCELED /
#   FILE_UPLOAD_URL_CREATED / FILE_UPLOAD_STARTED / FILE_UPLOAD_FAILED /
#   FILE_UPLOADED / FILE_CONFIRMED
# ⚠️ FILE_UPLOADED 语义：专指存储层确认字节已上传，禁止用于表示"upload url 已创建"。
# ⚠️ AI_REVIEW_STARTED 语义：Job PENDING→RUNNING 时使用，禁止复用 AI_REVIEW_ENQUEUED。
# ⚠️ FINAL_REVIEW_REQUESTED：DOUBLE_REVIEW 策略下人工审核通过时使用，非 REVIEW_ACCEPTED。
from sqlalchemy import Column, String, Text, DateTime, JSON, ForeignKey, func

from app.database import Base


class AuditLog(Base):
    __tablename__ = "audit_logs"

    # ID 由应用层生成，前缀 audit_，不使用数据库自增
    id = Column(String(64), primary_key=True, nullable=False)

    # 契约 AuditLog.entityType（最长 AI_REVIEW_JOB = 13 字符）：
    #   TASK / SCHEMA / ITEM / ASSIGNMENT / SUBMISSION / REVIEW / AI_REVIEW_JOB / EXPORT / FILE
    entity_type = Column(String(20), nullable=False)

    # 被审计的实体 ID
    entity_id = Column(String(64), nullable=False)

    # 契约 AuditAction，使用 String(40) 留有余量（最长值 25 字符）
    action = Column(String(40), nullable=False)

    # 契约 AuditLog.actor 中的 actor.id，FK → users.id
    actor_id = Column(String(64), ForeignKey("users.id"), nullable=False)

    # 契约 AuditLog.before?: unknown，操作前状态快照，可空
    before_json = Column(JSON, nullable=True)

    # 契约 AuditLog.after?: unknown，操作后状态快照，可空
    after_json = Column(JSON, nullable=True)

    # 契约 AuditLog.reason?: string，如打回原因、人工备注，可能较长使用 Text
    reason = Column(Text, nullable=True)

    # 契约 AuditLog.requestId?: string，请求追踪 ID
    request_id = Column(String(255), nullable=True)

    # 契约 AuditLog.createdAt；无 updated_at，本表记录写入后不可修改
    created_at = Column(DateTime, nullable=False, server_default=func.now())
